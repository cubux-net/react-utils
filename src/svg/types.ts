import { CSSProperties, FC, SVGProps } from 'react';

export type SvgFC = FC<SVGProps<SVGSVGElement>>;
export type CssTransform = NonNullable<CSSProperties['transform']>;
export type CalcTransform = (prev: CssTransform | undefined) => CssTransform;
