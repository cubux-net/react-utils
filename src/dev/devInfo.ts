import { ComponentType } from 'react';

export interface DevInfo {
  type: string;
  tag?: string;
  Orig?: ComponentType<any>;
  comment?: string;
}

let devInfo: WeakMap<ComponentType<any>, DevInfo>;
const _DEV = process.env.NODE_ENV !== 'production';
if (_DEV) {
  devInfo = new WeakMap();
}

export function setDevInfo<T extends DevInfo>(
  subject: ComponentType<any>,
  info: T,
) {
  if (_DEV) {
    devInfo!.set(subject, info);
  }
}

export function getDevInfo<T extends DevInfo = DevInfo>(
  subject: ComponentType<any>,
): T | undefined {
  if (_DEV) {
    return devInfo!.get(subject) as T | undefined;
  }
}
