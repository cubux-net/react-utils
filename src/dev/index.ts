export * from './deprecated';
export * from './devInfo';
