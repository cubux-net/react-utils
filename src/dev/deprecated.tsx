import React, {
  ComponentClass,
  ComponentProps,
  ComponentRef,
  ComponentType,
  forwardRef,
  ForwardRefExoticComponent,
  PropsWithoutRef,
  RefAttributes,
} from 'react';
import { DevInfo, setDevInfo } from './devInfo';

const _DEV = process.env.NODE_ENV !== 'production';

interface Options extends Omit<DevInfo, 'Orig' | 'type'> {
  /**
   * Whether to use `forwardRef()` for wrapped component
   *
   * When `false` (by default) the result component is just `FC`, so `ref={}`
   * React attribute cannot be used.
   *
   * Setting this property to `true` cause the result to be
   * `ForwardRefExoticComponent`.
   */
  withRef?: boolean;
}
interface OptionsWithRef extends Options {
  /** @inheritDoc */
  withRef: true;
}
interface OptionsWithoutRef extends Options {
  /** @inheritDoc */
  withRef?: false;
}

interface DeprecatedDecorator {
  /**
   * Wrap a component to emit deprecation warning
   *
   * A `getDevInfo()` can be used in development env to get details about
   * deprecation. This can be used on "dev only" internal pages.
   *
   * Does nothing in production env and returns origin component as is.
   *
   * ```tsx
   * function OldComponentOrigin() {
   *   return <div>...</div>;
   * }
   *
   * const OldComponent = deprecated(OldComponentOrigin, {
   *   comment: 'Use `NewComponent` instead.',
   * });
   * ```
   */
  <T extends ComponentClass<any>>(
    Orig: T,
    options: OptionsWithRef,
  ): ComponentType<
    PropsWithoutRef<ComponentProps<T>> & RefAttributes<ComponentRef<T>>
  >;

  /**
   * Wrap a component to emit deprecation warning
   *
   * A `getDevInfo()` can be used in development env to get details about
   * deprecation. This can be used on "dev only" internal pages.
   *
   * Does nothing in production env and returns origin component as is.
   *
   * ```tsx
   * function OldComponentOrigin() {
   *   return <div>...</div>;
   * }
   *
   * const OldComponent = deprecated(OldComponentOrigin, {
   *   comment: 'Use `NewComponent` instead.',
   * });
   * ```
   */
  <P, E>(
    Orig: ForwardRefExoticComponent<P & RefAttributes<E>>,
    options: OptionsWithRef,
  ): ComponentType<P & RefAttributes<E>>;

  /**
   * Wrap a component to emit deprecation warning
   *
   * A `getDevInfo()` can be used in development env to get details about
   * deprecation. This can be used on "dev only" internal pages.
   *
   * Does nothing in production env and returns origin component as is.
   *
   * ```tsx
   * function OldComponentOrigin() {
   *   return <div>...</div>;
   * }
   *
   * const OldComponent = deprecated(OldComponentOrigin, {
   *   comment: 'Use `NewComponent` instead.',
   * });
   * ```
   */
  <T extends ComponentType<any>>(
    Orig: T,
    options?: OptionsWithoutRef,
  ): ComponentType<PropsWithoutRef<ComponentProps<T>>>;
}

/**
 * Wrap a component to emit deprecation warning
 *
 * A `getDevInfo()` can be used in development env to get details about
 * deprecation. This can be used on "dev only" internal pages.
 *
 * Does nothing in production env and returns origin component as is.
 *
 * ```tsx
 * function OldComponentOrigin() {
 *   return <div>...</div>;
 * }
 *
 * const OldComponent = deprecated(OldComponentOrigin, {
 *   comment: 'Use `NewComponent` instead.',
 * });
 * ```
 */
export const deprecated: DeprecatedDecorator = _DEV
  ? <T extends ComponentType<any>>(
      Orig: T,
      { withRef, ...opt }: Options = {},
    ) => {
      function warn() {
        console.warn(
          'Deprecated component',
          Orig.displayName ?? Orig.name,
          Orig,
          'rendered.',
          opt.comment ?? '',
          new Error(),
        );
      }

      let C: ComponentType<ComponentProps<T>>;
      if (withRef) {
        const O: ComponentType<ComponentProps<T> & RefAttributes<any>> = Orig;
        C = forwardRef<any, ComponentProps<T>>((props, ref) => {
          warn();
          return <O {...props} ref={ref} />;
        }) as any;
      } else {
        C = (props: ComponentProps<T>) => {
          warn();
          return <Orig {...props} />;
        };
      }
      C.displayName = `${Orig.displayName ?? Orig.name}.deprecated${
        opt.tag ? `.${opt.tag}` : ''
      }`;
      setDevInfo(C, { ...opt, Orig, type: 'deprecated' });
      return C;
    }
  : <T extends ComponentType<any>>(Orig: T) => Orig;
