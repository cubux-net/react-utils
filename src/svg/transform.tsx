import React, { useMemo } from 'react';
import { setDevInfo } from '../dev';
import { CalcTransform, CssTransform, SvgFC } from './types';

const _DEV = process.env.NODE_ENV !== 'production';

/**
 * Creates new SVG FC applying a CSS `transform` to origin SVG FC.
 *
 * @param Orig Origin SVG FC to reuse
 * @param transform CSS transform to apply
 */
export function transform(
  Orig: SvgFC,
  transform: CssTransform | CalcTransform,
): SvgFC {
  let C: SvgFC;
  let tag: string;
  if (typeof transform === 'function') {
    C = ({ style, ...rest }) => (
      <Orig
        {...rest}
        style={useMemo(
          () => ({ ...style, transform: transform(style?.transform) }),
          [style],
        )}
      />
    );

    tag = 'function';
    if (_DEV) {
      setDevInfo(C, {
        Orig,
        type: 'transform',
        tag,
        comment: transform.toString(),
      });
    }
  } else {
    C = ({ style, ...rest }) => (
      <Orig
        {...rest}
        style={useMemo(
          () => ({
            ...style,
            transform:
              style?.transform !== undefined
                ? transform + ', ' + style.transform
                : transform,
          }),
          [style],
        )}
      />
    );

    tag = transform;
    if (_DEV) {
      setDevInfo(C, {
        Orig,
        type: 'transform',
        tag,
      });
    }
  }
  C.displayName = `${Orig.displayName ?? Orig.name}.transform{${tag}}`;
  return C;
}

/**
 * Creates new SVG FC applying horizontal flip to origin SVG FC with CSS
 * transform
 *
 * @param Orig Origin SVG FC to reuse
 */
export function flipH(Orig: SvgFC) {
  return transform(Orig, 'scaleX(-1)');
}

/**
 * Creates new SVG FC applying vertical flip to origin SVG FC with CSS transform
 *
 * @param Orig Origin SVG FC to reuse
 */
export function flipV(Orig: SvgFC) {
  return transform(Orig, 'scaleY(-1)');
}

/**
 * Creates new SVG FC applying rotation 90 deg anti-clockwise to origin SVG FC
 * with CSS transform
 *
 * @param Orig Origin SVG FC to reuse
 */
export function rot90L(Orig: SvgFC) {
  return transform(Orig, 'rotate(-90deg)');
}

/**
 * Creates new SVG FC applying rotation 90 deg clockwise to origin SVG FC with
 * CSS transform
 *
 * @param Orig Origin SVG FC to reuse
 */
export function rot90R(Orig: SvgFC) {
  return transform(Orig, 'rotate(90deg)');
}

/**
 * Creates new SVG FC applying rotation 180 deg to origin SVG FC with CSS transform
 *
 * @param Orig Origin SVG FC to reuse
 */
export function rot180(Orig: SvgFC) {
  return transform(Orig, 'rotate(180deg)');
}
