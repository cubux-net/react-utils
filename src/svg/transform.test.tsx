import '@testing-library/jest-dom';
import React from 'react';
import { render } from '@testing-library/react';
import { DevInfo, getDevInfo } from '../dev';
import { flipH, flipV, rot180, rot90L, rot90R, transform } from './transform';
import { SvgFC } from './types';

const SvgIcon: SvgFC = (props) => (
  <svg
    {...props}
    width={16}
    height={16}
    viewBox="0 0 16 16"
    fill="currentColor"
  >
    <path d="M2,5 h12 v6 z" />
  </svg>
);

describe('transform', () => {
  describe('string', () => {
    const ModifiedSvg = transform(SvgIcon, 'rotate(30deg)');

    expect(getDevInfo(ModifiedSvg)).toEqual<DevInfo>({
      Orig: SvgIcon,
      type: 'transform',
      tag: 'rotate(30deg)',
      comment: undefined,
    });

    expect(ModifiedSvg.displayName).toMatch(/\.transform\{rotate\(30deg\)}$/);

    it('default', () => {
      expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
    });
    it('extra', () => {
      expect(
        render(
          <ModifiedSvg
            version="1.0"
            style={{ transformOrigin: '40%, 40%', transform: 'scaleX(-1)' }}
          />,
        ).asFragment(),
      ).toMatchSnapshot();
    });
  });

  describe('callback', () => {
    const ModifiedSvg = transform(SvgIcon, (prev) =>
      prev ? 'rotate(30deg),' + prev : 'rotate(40deg)',
    );

    const info = getDevInfo(ModifiedSvg);
    expect(info).not.toBeUndefined();
    expect(info!.Orig).toBe(SvgIcon);
    expect(info!.type).toEqual('transform');
    expect(info!.tag).toEqual('function');
    expect(info!.comment).toMatch(
      /^function\b.*rotate\(30deg\),.*rotate\(40deg\)/s,
    );

    expect(ModifiedSvg.displayName).toMatch(/\.transform\{function}$/s);

    it('default', () => {
      expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
    });
    it('extra', () => {
      expect(
        render(
          <ModifiedSvg
            version="1.0"
            style={{ transformOrigin: '40%, 40%', transform: 'scaleX(-1)' }}
          />,
        ).asFragment(),
      ).toMatchSnapshot();
    });
  });
});

describe('flipH', () => {
  const ModifiedSvg = flipH(SvgIcon);

  it('default', () => {
    expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
  });
  it('extra', () => {
    expect(
      render(
        <ModifiedSvg
          version="1.0"
          style={{ transformOrigin: '40%, 40%', transform: 'rotate(30deg)' }}
        />,
      ).asFragment(),
    ).toMatchSnapshot();
  });
});

describe('flipV', () => {
  const ModifiedSvg = flipV(SvgIcon);

  it('default', () => {
    expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
  });
  it('extra', () => {
    expect(
      render(
        <ModifiedSvg
          version="1.0"
          style={{ transformOrigin: '40%, 40%', transform: 'rotate(30deg)' }}
        />,
      ).asFragment(),
    ).toMatchSnapshot();
  });
});

describe('rot90L', () => {
  const ModifiedSvg = rot90L(SvgIcon);

  it('default', () => {
    expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
  });
  it('extra', () => {
    expect(
      render(
        <ModifiedSvg
          version="1.0"
          style={{ transformOrigin: '40%, 40%', transform: 'rotate(30deg)' }}
        />,
      ).asFragment(),
    ).toMatchSnapshot();
  });
});

describe('rot90R', () => {
  const ModifiedSvg = rot90R(SvgIcon);

  it('default', () => {
    expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
  });
  it('extra', () => {
    expect(
      render(
        <ModifiedSvg
          version="1.0"
          style={{ transformOrigin: '40%, 40%', transform: 'rotate(30deg)' }}
        />,
      ).asFragment(),
    ).toMatchSnapshot();
  });
});

describe('rot180', () => {
  const ModifiedSvg = rot180(SvgIcon);

  it('default', () => {
    expect(render(<ModifiedSvg />).asFragment()).toMatchSnapshot();
  });
  it('extra', () => {
    expect(
      render(
        <ModifiedSvg
          version="1.0"
          style={{ transformOrigin: '40%, 40%', transform: 'rotate(30deg)' }}
        />,
      ).asFragment(),
    ).toMatchSnapshot();
  });
});
