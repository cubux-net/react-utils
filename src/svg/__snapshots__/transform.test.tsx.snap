// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flipH default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: scaleX(-1);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`flipH extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: scaleX(-1), rotate(30deg);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`flipV default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: scaleY(-1);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`flipV extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: scaleY(-1), rotate(30deg);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot90L default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: rotate(-90deg);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot90L extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: rotate(-90deg), rotate(30deg);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot90R default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: rotate(90deg);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot90R extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: rotate(90deg), rotate(30deg);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot180 default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: rotate(180deg);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`rot180 extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: rotate(180deg), rotate(30deg);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`transform callback default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: rotate(40deg);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`transform callback extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: rotate(30deg),scaleX(-1);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`transform string default 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform: rotate(30deg);"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;

exports[`transform string extra 1`] = `
<DocumentFragment>
  <svg
    fill="currentColor"
    height="16"
    style="transform-origin: 40%, 40%; transform: rotate(30deg), scaleX(-1);"
    version="1.0"
    viewBox="0 0 16 16"
    width="16"
  >
    <path
      d="M2,5 h12 v6 z"
    />
  </svg>
</DocumentFragment>
`;
